src/components/TodoFilter/TodoFilter.tsx

```diff
@@ -1,30 +1,89 @@
-          <option value="completed">Completed</option>
-        </select>
-      </span>
-    </p>
-
-    <p className="control is-expanded has-icons-left has-icons-right">
-      <input
-        data-cy="searchInput"
-        type="text"
-        className="input"
-        placeholder="Search..."
-      />
-      <span className="icon is-left">
-        <i className="fas fa-magnifying-glass" />
-      </span>
-
-      <span className="icon is-right" style={{ pointerEvents: 'all' }}>
-        {/* eslint-disable-next-line jsx-a11y/control-has-associated-label */}
-        <button data-cy="clearSearchButton" type="button" className="delete" />
-      </span>
-    </p>
-  </form>
-);
+import { useState, useEffect } from 'react';
+import { Todo } from '../../types/Todo';
+
+interface TodoFilterProps {
+  todos: Todo[];
+  onFilterChange: (filteredTodos: Todo[]) => void;
+}
+
+export enum FilterOptions {
+  ALL = 'all',
+  COMPLETED = 'completed',
+  ACTIVE = 'active',
+}
+
+export const TodoFilter: React.FC<TodoFilterProps> = ({
+  todos,
+  onFilterChange,
+}) => {
+  const [filter, setFilter] = useState<FilterOptions>(FilterOptions.ALL);
+  const [query, setQuery] = useState('');
+
+  const filterTodos = () => {
+    let filteredTodos = todos.filter(todo => {
+      switch (filter) {
+        case FilterOptions.ACTIVE:
+          return !todo.completed;
+        case FilterOptions.COMPLETED:
+          return todo.completed;
+        default:
+          return true;
+      }
+    });
+
+    if (query) {
+      filteredTodos = filteredTodos.filter(todo =>
+        todo.title.toLowerCase().includes(query.toLowerCase()),
+      );
+    }
+
+    onFilterChange(filteredTodos);
+  };
+
+  useEffect(() => {
+    filterTodos();
+  }, [filter, query, todos]);
+
+  return (
+    <form className="field has-addons">
+      <p className="control">
+        <span className="select">
+          <select
+            data-cy="statusSelect"
+            value={filter}
+            onChange={event => setFilter(event.target.value as FilterOptions)}
+          >
+            <option value="all">All</option>
+            <option value="active">Active</option>
+            <option value="completed">Completed</option>
+          </select>
+        </span>
+      </p>
+
+      <p className="control is-expanded has-icons-left has-icons-right">
+        <input
+          data-cy="searchInput"
+          type="text"
+          className="input"
+          placeholder="Search..."
+          value={query}
+          onChange={event => setQuery(event.target.value)}
+        />
+        <span className="icon is-left">
+          <i className="fas fa-magnifying-glass" />
+        </span>
+
+        {query && (
+          <span className="icon is-right" style={{ pointerEvents: 'all' }}>
+            <button
+              data-cy="clearSearchButton"
+              type="button"
+              className="delete"
+              onClick={() => setQuery('')}
```

Comment:
You are generating a new function on every render. According to the checklist, you should avoid generating keys or functions on render. Consider using a callback function with `useCallback` hook.